dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/authorize/AbstractScopeAuthorizationStrategy.java

```diff
@@ -0,0 +1,113 @@
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import lombok.NonNull;
+import lombok.val;
+import lombok.extern.slf4j.Slf4j;
+
+import org.icgc.dcc.storage.server.service.MetadataService;
+import org.icgc.dcc.storage.server.util.AuthScope;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+
+@Slf4j
+public abstract class AbstractScopeAuthorizationStrategy {
+
+  protected AuthScope scope;
+
+  @Autowired
+  protected MetadataService metadataClient;
+
+  public AbstractScopeAuthorizationStrategy() {
+    super();
+  }
+
+  /**
+   * Package-private constructor for use in unit tests (to initialize scope context) without a Spring Context
+   */
+  protected AbstractScopeAuthorizationStrategy(final String scopeStr) {
+    setAuthorizeScope(scopeStr);
+    scope = AuthScope.from(getAuthorizeScope());
+  }
+
+  protected abstract void setAuthorizeScope(String scopeStr);
+
+  protected abstract String getAuthorizeScope();
+
+  protected abstract boolean verify(@NonNull List<AuthScope> grantedScopes, @NonNull final String objectId);
+
+  /**
+   * Main entry point: version of authorize() method that assumes object id is passed-in (method-level security has
+   * access to this)
+   */
+  public boolean authorize(@NonNull Authentication authentication, @NonNull final String objectId) {
+    log.info("Checking authorization with object id {}", objectId);
+    scope = AuthScope.from(getAuthorizeScope());
+
+    // TODO: urrrrr
+    // if not OAuth2, then no scopes available at all
+    List<AuthScope> grantedScopes = Collections.<AuthScope> emptyList();
+    if (authentication instanceof OAuth2Authentication) {
+      OAuth2Authentication o2auth = (OAuth2Authentication) authentication;
+      grantedScopes = getScopes(o2auth);
+    }
+
+    return verify(grantedScopes, objectId);
+  }
+
+  /**
+   * Handles OAuth2Authentication object
+   * @param o2auth from Authentication token
+   * @return collection of AuthScope objects
+   */
+  protected List<AuthScope> getScopes(@NonNull OAuth2Authentication o2auth) {
+    val scopeStrs = o2auth.getOAuth2Request().getScope();
+    return extractScopes(scopeStrs);
+  }
+
+  /**
+   * Filters out any AuthScope instances that don't match the active system/operation.
+   */
+  protected List<AuthScope> extractScopes(@NonNull Set<String> scopeStrs) {
+    val result = scopeStrs.stream().map(s -> AuthScope.from(s))
+        .filter(p -> p.matches(scope))
+        .collect(Collectors.toList());
+    return result;
+  }
+
+  protected boolean validate(@NonNull String uuid5) {
+    try {
+      UUID.fromString(uuid5);
+      return true;
+    } catch (IllegalArgumentException e) {
+      return false;
+    }
+  }
+
+  void setMetadataClient(MetadataService mock) {
+    metadataClient = mock;
```

Comment:
The "Service" is actually the client of the Metadata Server. It's the "service" pattern rather than a server service.